### ISSUE TITLE
Semantics of SplitText carrier

### ISSUE THREAD

**Issue by yurishkuro** [opentracing/basictracer-go#7]
@bensigelman @tschottdorf I was looking at the implementation of split-text codecs in the basictracer, and noticed a big difference from our own implementation, specifically in how it handles the baggage items by simply copying the internal k/v map of the span into the `carrier.Baggage` map. The way I always imagined using the split-text carrier was:

```go
func call_out(span Span, request Request) {
    // Imagine Request represents the headers as a map[string]string (unlike Go http.Request).
    carrier := &SplitTextCarrier { TracerState: request.headers, Baggage: request.headers }
    span.Tracer().Inject(span, SpitTextFormat, carrier)
    execute(request)
}
```
Notice that we can use the headers map directly as the target of injection. It's also quite possible that the same map already contained some other application- or protocol-specific headers. Which means when we decode a similar map back into the Span, the end-user code would also use a single map

```go
func handle_incoming(request Request) {
    carrier := &SplitTextCarrier { TracerState: request.headers, Baggage: request.headers }
    span, _ := tracer.Join("op name", SpitTextFormat, carrier)
    ...
}
```

And that's where the problem with basictracer implementation comes in - it expects that Carrier.Baggage is the exact map of baggage items with exact key names provided by the end user originally. I don't see how it can work in an RPC protocol that supports custom application headers, unless the RPC frameworks takes an additional step of encoding keys/values of baggage items after they are injected into carrier (which is non-starter since there may be a different RPC framework on the receiving end).

In contrast, our previous implementation was always using a fixed prefix when converting baggage key into text carrier key. The use of the prefix helps find the baggage items on the receiving side by scanning through all app headers and looking for keys that start with that prefix.

What are your thoughts?  I feel that we need to clarify the semantics of the SplitTextCarrier a bit more, as at least in my mind there were more expectations about it than what basictracer has implemented.

**Comment by tschottdorf** [comment]
In `Go`, there is the `SplitTextCarrier` but also the `GoHTTPHeaderCarrier` (or whatever the name). I would expect the latter to "reserve" its subspace in the headers through, say, a prefix. Similarly for python. But at the end of the day, the standard types need to be useful - I'm not personally invested in `SplitText` carrier, but it looks like a HeaderCarrier which behaves how you describe above should be a standard.

**Comment by yurishkuro** [comment]
I think of SplitTextCarrier as HeaderCarrier already. I don't know where the behavior as implemented in basictracer would be useful - the _only_ time it can work is if you have an RPC framework that is identical in both server and client, so that it can manually disambiguate the keys in SplitTextCarrier.baggage. That does not sound like a particular useful use case to me.

On the other hand GoHTTPHeaderCarrier is very special use case. It cannot be generalized across different languages, which is what I'm after - a recipe for working with a split text carrier regardless of the language.

**Comment by yurishkuro** [comment]
To emphasize/rephrase: GoHTTPHeaderCarrier is a very special use case, won't always work even in Go if you use non-http RPC framework. But a string->string map is an absolutely universal use case, in  every language and most RPC frameworks.

**Comment by bensigelman** [comment]
@yurishkuro SplitTextCarrier is tricky/subtle to program against for the reasons we're discussing here... I noticed essentially what you're bringing up and mentioned it [here](https://github.com/opentracing/basictracer-go/pull/4#discussion-diff-54029985). We could change the SplitTextCarrier API to do some of the prefix addition/removal internally (perhaps my turning it into a Go `interface`). Thoughts on that approach?

TL;DR, it would be nice if users of SplitTextCarrier (including `basictracer-go`) could ignore the issue of prefixes entirely and build a little more functionality into the carrier proper. For what it's worth, I really liked @tschottdorf's DelegatingCarrier and can imagine something similar for SplitText specifically.

**Comment by yurishkuro** [comment]
@bensigelman had a look at that PR #4, but don't see yet how the end user can be shielded from knowing what exact header names are used by the tracer, even with Delegating Carrier (which is somewhat similar to a "streaming carrier" I proposed some time ago).

I think the question I post here is different from whether the carrier has explicit data structure (a map) or a map-like delegation methods. The problem is, as you said, that the carrier _always_ represents a superset of key/value pairs. A carrier is essentially a facade over RPC request. The tracer injects a subset of keys, and extracts a subset of keys from that RPC request's "headers". That is what I want to clarify in the spec. Whether the tracer uses a prefix or a single key for all of the baggage is implementation detail. The basic tracer's implementation imo is not compliant with that expectation.

**Comment by bensigelman** [comment]
I'm just typing this out most hackily so apologies if it doesn't make sense:

```
type SplitTextCarrier interface {
    SetTracerState(textMap map[string]string)
    SetBaggage(textMap map[string]string)

    // This would add prefixes for the SplitTextCarrier user. See
    // NewSplitTextCarrierFromHostMap().
    EmbedInHostMap(hostMap map[string]string)

    // Could be left out of the API entirely, and should certainly have
    // reasonable defaults:
    SetTracerStatePrefix(p string)
    SetBaggagePrefix(p string)
}

func NewSplitTextCarrierFromHostMap(hostMap map[string]string) SplitTextCarrier { ... }
```

^^^ again, not fully baked, but just trying to convey how the carrier could possibly handle the prefixes itself?

**Comment by yurishkuro** [comment]
@bensigelman yeah, you may need to clarify, I am not really getting what you mean. 

I don't think SplitTextCarrier requires any API changes. What it requires is a semantic definition of the maps inside it being supersets of the tracer state and baggage. Tracer can only read from the maps what it wrote, by whatever means necessary (a prefix, or a single key, or another clever trick). And the following unit test:

```go
sp1 := tracer.StartSpan("s1")
sp1.SetBaggageItem("x", "1")
m := map[string]string {"y": "2"}
c := &SplitTextCarrier{m, m}
tracer.Inject(sp1, SplitText, c)
assert.Equal(t, "2", m["y"], "key 'y' is still in the map, which is a superset")
// reverse
sp2, _ := tracer.Join("s2", SplitText, c)
assert.Equal(t, "1", sp2.BaggageItem("x"), "key 'x' restored as baggage")
assert.Equal(t, "", sp2.BaggageItem("y"), "key 'y' not in baggage")
```

The last assert is the critical part, which is broken by the current basictracer implementation.

**Comment by bensigelman** [comment]
Something about this arrangement doesn't quite smell right to me. I think it has to do with the fact that we always pass the same map to both items in the "split" (sic) tuple... that redundancy suggests a design problem. As an alternative, imagine if it were called `MapSupersetCarrier` and only had one internal `map[string]string` – IMO that would be clearer and would still work well with HTTP headers. It could also expose an `EnableBaggage bool` (or similar) to give the user veto power over baggage in general.

```go
struct SupersetCarrier {
    LargerMap map[string]string
    EnableBaggage bool
}

...

c := &SupersetCarrier{req.Header, true}
tracer.Inject(sp1, Superset, c)
...
sp2, _ := tracer.Join("s2", Superset, c)
```

Thoughts? Upon reflection, it just seems weird to me to have these split maps in SplitTextCarrier that need prefixes/etc to differentiate fields. Either my brain wants the maps to be consolidated or for the prefix requirement to go away as far as the OT user is concerned.

(Note that we *can* do something morally similar to what I was proposing with `EmbedInHostMap` earlier, but I'd rather not take the time to more properly explain that approach since I don't like where it's headed anyway)

**Comment by yurishkuro** [comment]
I don't think the "superset" concept is directly related to one vs. two maps. I can modify my unit test to use `c := &SplitTextCarrier{m1, m2}`, and it would still fail in `basictracer`. I am not suggesting a concept of a "prefix" to be exposed to the end user - a prefix is internal implementation detail of the tracer (if it wants it to be). The main conditions I am proposing are:
 1. a map in the carrier does not have to be empty when given to injector;
 1. a map may contain keys unrelated to tracer-state/baggage when given to extractor. 

It is the tracer's responsibility to figure out which keys from the map are relevant for extraction. Possible implementations of that responsibility are (1) use only a single slot for baggage (how basictracer implemented GoHeader carrier), or (2) use a tracer-preferred prefix (how our tracer implements it).

**Comment by bensigelman** [comment]
To be clear, I just think it's *clearer* in the "superset" case. SplitTextCarrier can be made to work, but it's going to require some really careful comments since there are some real subtleties here. The conditions you're proposing make sense to me, assuming we plow forward with SplitTextCarrier (though I am not immune to arguments to get rid of it and replace with something that will be more intuitive).

**Comment by bg451** [comment]
Closing this now that github.com/opentracing/basictracer-go/pull/15 is in effect.
